### PR TITLE
Fix BigQuery configuration with parent-project-id

### PR DESCRIPTION
**Pull request by pajaks**
## Description

If `bigquery.parent-project-id` was specified in configuration it was set as `projectId` in `BigQueryOptions`. 
For queries without `TableId` or `DatasetId` specified `BigQueryOptions` was used resulting in using wrong projectId.

This problem is visible when using PTF with parent-proejct-id set, resulting in permission problem like:
```
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
POST https://bigquery.googleapis.com/bigquery/v2/projects/parent-project-id/jobs?prettyPrint=false
{
  "code": 403,
  "errors": [
    {
      "domain": "global",
      "message": "Access Denied: Table parent-project-id:tpch.nation: User does not have permission to query table parent-project-id:tpch.nation, or perhaps it does not exist.",
      "reason": "accessDenied"
    }
  ],
  "message": "Access Denied: Table parent-project-id:tpch.nation: User does not have permission to query table parent-project-id:tpch.nation, or perhaps it does not exist.",
  "status": "PERMISSION_DENIED"
}
```

Billing project id should be set using `setQuotaProjectId'  according to https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.ServiceOptions#com_google_cloud_ServiceOptions_getQuotaProjectId__

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# BigQuery
* Fix configuration with parent-project-id. ({issue}`23041`)
```

### CODE REVIEW

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/CredentialsOptionsConfigurer.java [hunk 1]
```diff
@@ -31,22 +31,26 @@ public class CredentialsOptionsConfigurer
...
-    private final Optional<String> parentProjectId;
+    private final Optional<String> configProjectId;
+    private final Optional<String> configParentProjectId;
 
     @Inject
     public CredentialsOptionsConfigurer(BigQueryConfig bigQueryConfig, BigQueryCredentialsSupplier credentialsSupplier)
     {
-        this.parentProjectId = requireNonNull(bigQueryConfig, "bigQueryConfig is null").getParentProjectId();
+        this.configProjectId = requireNonNull(bigQueryConfig, "bigQueryConfig is null").getProjectId();
+        this.configParentProjectId = requireNonNull(bigQueryConfig, "bigQueryConfig is null").getParentProjectId();
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/CredentialsOptionsConfigurer.java:41]: `requireNonNull` for config class is redundant. See https://github.com/trinodb/trino/pull/13940 for more details.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 2]
```diff
@@ -1077,6 +1083,25 @@ public void testNativeQueryIncorrectSyntax()
                 .failure().hasMessageContaining("Failed to get destination table for query");
     }
 
+    @Test
+    public void testQueriesWithParentProjectId()
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:1087]: Can we add a new test class? e.g. `TestBigQueryParentProjectId`

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 3]
```diff
@@ -1077,6 +1083,25 @@ public void testNativeQueryIncorrectSyntax()
...
+            throws Exception
+    {
+        DistributedQueryRunner queryRunner = BigQueryQueryRunner.builder()
+                .setConnectorProperties(ImmutableMap.<String, String>builder()
+                        .put("bigquery.arrow-serialization.enabled", "false")
+                        .put("bigquery.job.label-name", "trino_query")
+                        .put("bigquery.job.label-format", "q_$QUERY_ID__t_$TRACE_TOKEN")
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:1094]: These config properties look unncessary for testing a parent project id.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 4]
```diff
@@ -1077,6 +1083,25 @@ public void testNativeQueryIncorrectSyntax()
...
+                        .put("bigquery.project-id", testingProjectId)
+                        .put("bigquery.parent-project-id", testingParentProjectId)
+                        .buildOrThrow())
+                .build();
+
+        assertThat(queryRunner.execute(getSession(), "SELECT name FROM bigquery.tpch.nation WHERE nationkey = 0").getOnlyValue()).isEqualTo("ALGERIA");
+        assertThat(queryRunner.execute(getSession(), "SELECT * FROM TABLE(bigquery.system.query(query => 'SELECT name FROM tpch.nation WHERE nationkey = 0'))").getOnlyValue()).isEqualTo("ALGERIA");
+        assertThat(queryRunner.execute(getSession(), format("SELECT * FROM TABLE(bigquery.system.query(query => 'SELECT name FROM %s.tpch.nation WHERE nationkey = 0'))", testingProjectId)).getOnlyValue()).isEqualTo("ALGERIA");
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:1102]: Can we use `computeScalar` method?

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 5]
```diff
@@ -1077,6 +1083,25 @@ public void testNativeQueryIncorrectSyntax()
                 .failure().hasMessageContaining("Failed to get destination table for query");
     }
 
+    @Test
+    public void testQueriesWithParentProjectId()
+            throws Exception
+    {
+        DistributedQueryRunner queryRunner = BigQueryQueryRunner.builder()
+                .setConnectorProperties(ImmutableMap.<String, String>builder()
+                        .put("bigquery.arrow-serialization.enabled", "false")
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:1092]: Why do we need to disable this property?

Comment (pajaks) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:1092]: It fails without it:
```
1) Error: BigQuery connector requires additional JVM arguments to run when 'bigquery.arrow-serialization.enabled' is enabled. Please add '--add-opens=java.base/java.nio=ALL-UNNAMED' to the JVM configuration.
  at BigQueryConnectorModule$ClientModule.verifyPackageAccessAllowed(BigQueryConnectorModule.java:154)
      \_ installed by: BigQueryConnectorModule -> BigQueryConnectorModule$ClientModule -> ConditionalModule -> BigQueryConnectorModule$ClientModule$$Lambda/0x00000fff00b4ba78
```

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 4 above)
Comment (pajaks) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: After introducing new test class I followed `assertQuery()` convention.

File: plugin/trino-bigquery/pom.xml [hunk 6]
```diff
@@ -537,6 +537,7 @@
                                 <exclude>**/TestBigQueryCaseInsensitiveMapping.java</exclude>
                                 <exclude>**/TestBigQuery*FailureRecoveryTest.java</exclude>
                                 <exclude>**/TestBigQueryWithProxyTest.java</exclude>
+                                <exclude>**/TestBigQueryParentProjectId.java</exclude>
```
Comment (SemionPar) [inline plugin/trino-bigquery/pom.xml:540]: *question: * is this test just temporary excluded (until we have infra in place)?

Comment (pajaks) [reply plugin/trino-bigquery/pom.xml:540]: Forgot to include in cloud-1 profile. Fixed.

**Comment by findinpath** [comment]
```
# BigQuery
* Fix configruation with parent-project-id. ({issue}`23041`)
```

nit: typo in `configruation`

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -639,11 +639,15 @@ jobs:
...
+          TESTING_PROJECT_ID: ${{ vars.TESTING_PROJECT_ID }}
+          TESTING_PARENT_PROJECT_ID: ${{ vars.TESTING_PARENT_PROJECT_ID }}
         if: matrix.modules == 'plugin/trino-bigquery' && !contains(matrix.profile, 'cloud-tests-2') && (env.CI_SKIP_SECRETS_PRESENCE_CHECKS != '' || env.BIGQUERY_CREDENTIALS_KEY != '')
         run: |
           $MAVEN test ${MAVEN_TEST} -pl :trino-bigquery -Pcloud-tests-1 \
             -Dbigquery.credentials-key="${BIGQUERY_CREDENTIALS_KEY}" \
-            -Dtesting.gcp-storage-bucket="${GCP_STORAGE_BUCKET}"
+            -Dtesting.gcp-storage-bucket="${GCP_STORAGE_BUCKET}" \
+            -Dtesting.project-id="${BIGQUERY_TESTING_PROJECT_ID}" \
+            -Dtesting.parent-project-id="${BIGQUERY_TESTING_PARENT_PROJECT_ID}"
```
Comment (ebyhr) [inline .github/workflows/ci.yml:650]: These JVM options' naming are too generic. Please append `bigquery-`.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java [hunk 8]
```diff
@@ -259,7 +259,7 @@ public TableInfo getCachedTable(Duration viewExpiration, TableInfo remoteTableId
      */
     public String getParentProjectId()
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java:260]: Please update the method name. e.g. `getQuotaProjectId`

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import com.google.common.collect.ImmutableMap;
+import io.trino.testing.AbstractTestQueries;
+import io.trino.testing.QueryRunner;
+import org.junit.jupiter.api.Test;
+
+import static io.trino.testing.TestingProperties.requiredNonEmptySystemProperty;
+import static io.trino.tpch.TpchTable.NATION;
+import static java.lang.String.format;
+
+public class TestBigQueryParentProjectId
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java:26]: https://trino.io/docs/current/develop/tests.html#conventions-and-recommendations
> Test classes should be defined as package-private and final.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                .setConnectorProperties(ImmutableMap.<String, String>builder()
+                        .put("bigquery.project-id", testingProjectId)
+                        .put("bigquery.parent-project-id", testingParentProjectId)
+                        .buildOrThrow())
+                .setInitialTables(ImmutableList.of(NATION))
+                .build();
+    }
+
+    @Test
+    public void testQueriesWithParentProjectId()
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java:47]: https://trino.io/docs/current/develop/tests.html#conventions-and-recommendations
> Test methods should be defined as package-private.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        assertQuery(
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java:49]: `assertQuery` is deprecated though the method doesn't have the annotation. Please use `assertThat()` method instead.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.jupiter.api.Test;
+
+import static io.trino.testing.TestingProperties.requiredNonEmptySystemProperty;
+import static io.trino.tpch.TpchTable.NATION;
+import static java.lang.String.format;
+
+public class TestBigQueryParentProjectId
+        extends AbstractTestQueries
+{
+    private String testingProjectId;
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java:29]: We could initialize this variable in a constructor and make it final.

**Review by ebyhr — COMMENTED** [review]
Please squash 4th and 3rd commits into one.

**Comment by ebyhr** [comment]
/test-with-secrets sha=56ec6f35fda5235416262aab2120a25bdaf304ca

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java (hunk 8 above)
Comment (pajaks) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java:260]: I updated method so now it does not always return `QuotaProjectId`, but projectId dependent on configuration. Name is aligned with CreateReadSessionRequest API:
https://github.com/trinodb/trino/blob/e13da7386edd1c55e984971664c6b33deb583172/plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ReadSessionCreator.java#L103-L104

**Comment by pajaks** [comment]
Big Query tests with secretes passed in mirror PR https://github.com/trinodb/trino/pull/23141
Other tests failures not related to changes

**Review by hashhar — COMMENTED** [review]
It would be very helpful if we can have a short comment somewhere explaining the differences between project id, parent project id and how those are related to the quota/billing project.

Code itself seems fine but the semantics are unclear to me.

**Comment by pajaks** [comment]
> It would be very helpful if we can have a short comment somewhere explaining the differences between project id, parent project id and how those are related to the quota/billing project.

Quota project id is different name for parent project id, both indicates project used for quota and billing purposes.
Added comment in code.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+class TestBigQueryParentProjectId
+        extends AbstractTestQueryFramework
+{
+    private final String testingParentProjectId;
+    private final String testingProjectId;
+
+    TestBigQueryParentProjectId()
+    {
+        testingProjectId = requiredNonEmptySystemProperty("bigquery-testing.project-id");
+        testingParentProjectId = requiredNonEmptySystemProperty("bigquery-testing.parent-project-id");
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryParentProjectId.java:36]: nit: I prefer using the same order between fields and the constructor's initialization.

**Comment by ebyhr** [comment]
> Fix configuration with parent-project-id

@pajaks Thanks for writing the release note entry. I think users can't understand what was the issue from this sentence. Could you update it? We usually use "Fix failure when ..." for bug fix entries.

**Pull request merged**